### PR TITLE
add multi-sig-transfer

### LINKED ISSUE #210: WalletAPI.Transfer does not work for sending from genesis account
Attempting to send 1000 neo from genesis account to a standard address using a single node Neo3 chain (master branch). The consensus node wallet has two accounts, the node-specific account (`NU78azX2nqbuDPthu5UGMaCXTVwpAkpvye`) and the multi-sig account (`NfC9EW6UJEbfX4TKCnqGf72AKHXUfR71ma`). I am calling GetKey() on the multi-sig WalletAccount and passing it to WalletApi.Transfer. Transfer throws an exception `Insufficient GAS in address: NU78azX2nqbuDPthu5UGMaCXTVwpAkpvye`. That is the node-specific account address, NOT the multi-sig account address.

Stepping thru the transfer function, the exception is throwing in the TransactionManager.Sign() function. Sign is checking the gas balance, finding the gas balance to be zero and throwing.

### LINKED ISSUE #211: Utility.ToBigInteger should be public

### LINKED ISSUE #223: TransactionManager.MakeTransaction throws exception
Attempting to implement [transfer from multi-sig account](https://github.com/neo-project/docs/blob/release-neo3/docs/en-us/tooldev/sdk/transaction.md#constructing-a-transaction-to-transfer-from-multi-signature-account). TransactionManager.MakeTransaction throws an exception: `Exception thrown: 'Neo.Network.RPC.RpcException' in RpcClient.dll: 'Unable to cast object of type 'Neo.IO.Json.JString' to type 'Neo.IO.Json.JArray'.'`

It appears that RpcClient.InvokeScript is encoding the parameters incorrectly.  [RpcServer `invokescript`](https://github.com/neo-project/neo-modules/blob/d62b18bcea99186614c3043603fcaa8e6214ba77/src/RpcServer/RpcServer.SmartContract.cs#L89) looks like it's expecting CheckWitnessHashes to be an array of it's own while RpcClient.InvokeScript is simply appending the witness hashes to the parameter array

### PR DESCRIPTION
**Pull request by chenquanyu**
fix #211 
fix  #210 
fix #223

### CODE REVIEW

File: src/RpcClient/TransactionManager.cs [hunk 1]
```diff
@@ -84,8 +84,6 @@ public TransactionManager MakeTransaction(byte[] script, TransactionAttribute[]
                 long remainder = Tx.SystemFee % d;
                 if (remainder > 0)
                     Tx.SystemFee += d - remainder;
-                else if (remainder < 0)
```
Comment (shargon) [inline src/RpcClient/TransactionManager.cs:87]: Why this change?

Comment (chenquanyu) [reply src/RpcClient/TransactionManager.cs:87]: image
This case can never be met for the `Tx.SystemFee > 0`.

File: src/RpcClient/WalletAPI.cs [hunk 2]
```diff
@@ -156,6 +157,23 @@ public Transaction Transfer(UInt160 scriptHash, KeyPair from, UInt160 to, BigInt
...
+        /// Transfer NEP5 token from multi-sig account
+        /// </summary>
+        /// <param name="scriptHash">contract script hash</param>
+        /// <param name="m">multi-sig min signature count</param>
+        /// <param name="pubKeys">multi-sig pubKeys</param>
+        /// <param name="keys">sign keys</param>
+        /// <param name="to">to account</param>
+        /// <param name="amountInteger">transfer amount</param>
+        /// <returns></returns>
+        public Transaction Transfer(UInt160 scriptHash, int m, ECPoint[] pubKeys, KeyPair[] keys, UInt160 to, BigInteger amountInteger)
```
[the remark this answers is not in the document]
Comment (chenquanyu) [reply src/RpcClient/WalletAPI.cs:170]: Yes, we may wrap async methods for code reuse. https://github.com/chenquanyu/neo-modules/blob/add-async-methods/src/RpcClient/RpcClient.Async.cs

File: src/RpcClient/TransactionManager.cs [hunk 3]
```diff
@@ -150,6 +148,21 @@ public TransactionManager AddMultiSig(KeyPair key, int m, params ECPoint[] publi
...
+        /// Add Multi-Signature
+        /// </summary>
+        /// <param name="keys">The KeyPairs to sign transction</param>
+        /// <param name="m">The least count of signatures needed for multiple signature contract</param>
+        /// <param name="publicKeys">The Public Keys construct the multiple signature contract</param>
+        public TransactionManager AddMultiSig(KeyPair[] keys, int m, params ECPoint[] publicKeys)
+        {
+            foreach (var key in keys)
+            {
+                AddMultiSig(key, m, publicKeys);
```
Comment (shargon) [inline src/RpcClient/TransactionManager.cs:161]: We should avoid the line 146 (https://github.com/neo-project/neo-modules/pull/213/files#diff-b0495f29c2f433fb99528ffde111175eR146), I think that it's better to duplicate the logic here.

File: src/RpcClient/Nep5API.cs [hunk 4]
```diff
@@ -117,5 +118,32 @@ public Transaction CreateTransferTx(UInt160 scriptHash, KeyPair fromKey, UInt160
...
+        /// <param name="scriptHash">contract script hash</param>
+        /// <param name="m">multi-sig min signature count</param>
+        /// <param name="pubKeys">multi-sig pubKeys</param>
+        /// <param name="fromKeys">sign keys</param>
+        /// <param name="to">to account</param>
+        /// <param name="amount">transfer amount</param>
+        /// <returns></returns>
+        public Transaction CreateTransferTx(UInt160 scriptHash, int m, ECPoint[] pubKeys, KeyPair[] fromKeys, UInt160 to, BigInteger amount)
+        {
+            if (m > fromKeys.Length) throw new System.Exception($"Need at least {m} KeyPairs for signing!");
```
Comment (shargon) [inline src/RpcClient/Nep5API.cs:134]: || m<0 ?

Comment (chenquanyu) [reply src/RpcClient/Nep5API.cs:134]: Maybe this is also an ArgumentExcception.

image

File: src/RpcClient/TransactionManager.cs (hunk 3 above)
Comment (chenquanyu) [reply src/RpcClient/TransactionManager.cs:161]: More keys can be added after this method, so I didn't check keys count in this method.

File: src/RpcClient/TransactionManager.cs [hunk 5]
```diff
@@ -84,8 +84,6 @@ public TransactionManager MakeTransaction(byte[] script, TransactionAttribute[]
                 long remainder = Tx.SystemFee % d;
                 if (remainder > 0)
                     Tx.SystemFee += d - remainder;
```
Comment (shargon) [inline src/RpcClient/TransactionManager.cs:86]: I think that now we can remove the `remainder`

Comment (chenquanyu) [reply src/RpcClient/TransactionManager.cs]: Done by #219

File: src/RpcClient/WalletAPI.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (devhawk) [reply src/RpcClient/WalletAPI.cs:170]: The underlying HttpClient doesn't expose sync versions of Get, Post, etc since Http requests are inherently async. Shouldn't all the RpcWrapper methods follow the same strategy? Why are there sync methods for any of these RPC wrapper calls?

[the remark this answers is not in the document]
Comment (chenquanyu) [reply src/RpcClient/WalletAPI.cs:170]: I think we can add async methods after the discussion in #209.

**Comment by ProDog** [comment]
There are some exceptions when `SendRawTransaction`, maybe need update `CreateTransferTx`?

image

**Comment by chenquanyu** [comment]
Tested on the preview2 version.
image

**Comment by chenquanyu** [comment]
The signature verification failed after https://github.com/neo-project/neo/commit/bf16df9d5a090dfcaff90d7c9d5454cf5ad09101

**Comment by ProDog** [comment]
> The signature verification failed after [neo-project/neo@bf16df9](https://github.com/neo-project/neo/commit/bf16df9d5a090dfcaff90d7c9d5454cf5ad09101)

Only use default `"Magic": 5195086` can work.

**Comment by shargon** [comment]
Did you use the same `protocol.json` for sign?

Default `Magic` it's loaded from this file

https://github.com/neo-project/neo/commit/bf16df9d5a090dfcaff90d7c9d5454cf5ad09101#diff-dd6fe58934e6f742301fccc0758b474fR13

https://github.com/neo-project/neo/blob/bf16df9d5a090dfcaff90d7c9d5454cf5ad09101/src/neo/ProtocolSettings.cs#L36

**Comment by ProDog** [comment]
RpcClient always uses the default `Magic` by reference, but test net or private net not necessarily used default `Magic`.

**Comment by shargon** [comment]
You need to test it with the same `protocol.json` of the remote server

**Comment by ProDog** [comment]
> You need to test it with the same `protocol.json` of the remote server

Also need `protocol.json`  in RpcClient? :worried:

**Comment by shargon** [comment]
Yes, because the signature must be network dependant.

**Comment by ProDog** [comment]
RpcServer's `invokefunction` has the same problem, how to modify its `Magic` :thinking:

**Comment by chenquanyu** [comment]
RpcServer use the settings of CLI @ProDog
